lib/optimized/coin_flip_fix.dart

```diff
@@ -0,0 +1,233 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/scheduler.dart';
+import 'package:intl/intl.dart';
+import 'dart:math';
+
+final Random random = Random();
+final NumberFormat formatter = NumberFormat.percentPattern('en_US');
+
+class CoinFlipFix extends StatefulWidget {
+  @override
+  CoinFlipFixState createState() => CoinFlipFixState();
+}
+
+class CoinFlipFixState extends State<CoinFlipFix> with SingleTickerProviderStateMixin {
+  AnimationController _controller;
+  Animation _flipAnimation;
+  SchedulerBinding _scheduler;
+
+  // The total number of tasks we will run. We need to split up the computational
+  // work of calculating coin flips into multiple tasks. Each task should complete
+  // in <= 4ms. This will prevent UI jank and dropped frames from occurring due
+  // to a blocked UI thread.
+  static final int _totalTasks = 50;
+  // The current running task number. Will be in the range of 1 to [_totalTasks]
+  // (inclusive).
+  int _taskNumber = 1;
+
+  // Whether the coin is currently flipping
+  bool isCoinFlipping = false;
+  // The number of times the coin flip resulted in heads.
+  int _headsCount = 0;
+  // The number of times the coin flip resulted in tails.
+  int _tailsCount = 0;
+  // The total number of times the coin has been flipped.
+  int _totalFlipCount = 0;
+
+  void _flipCoin() {
+    if (!isCoinFlipping) {
+      assert(_taskNumber == 1);
+      _headsCount = 0;
+      _tailsCount = 0;
+      _totalFlipCount = 250000;
+      isCoinFlipping = true;
+
+      _maybeScheduleTask();
+    }
+  }
+
+  void _maybeScheduleTask() {
+    if (_taskNumber <= _totalTasks) {
+      // For each task, do a small chunk of work that will complete in <= 4ms:
+      // _totalFlipCount / _totalTasks = 5000 flips per task.
+      _scheduler.scheduleTask(() => _flipNTimes(5000), Priority.animation);
+    } else {
+      // At this point, all of the tasks have been completed. Update the UI.
+      setState(() {
+        _taskNumber = 1;
+        isCoinFlipping = false;
+      });
+    }
+  }
+
+  void _flipNTimes(int n) {
+    for (int i = 0; i < n; i++) {
+      bool isHeads = random.nextInt(2) % 2 == 0;
+      if (isHeads) {
+        _headsCount++;
+      } else {
+        _tailsCount++;
+      }
+    }
+    _taskNumber++;
```

Comment:
perhaps an easier pattern for users to understand is if instead of tracking task #, you track the amount of work done. So instead track
_totalCoinFlips
and schedule more flips when the _totalCoinFlips is less than the _goalCoinFlips.
Alternately track coinFlipsLeft
and end when coinFlipsLeft reaches zero.

Reply:
Done. Tracking total flips completed.